evalml/data_checks/datetime_format_data_check.py

```diff
@@ -118,21 +179,66 @@ def validate(self, X, y):
-                if self.datetime_column != "index"
-                else "either index"
-            )
+        col_name = (
+            self.datetime_column if self.datetime_column != "index" else "either index"
+        )
+
+        is_increasing = pd.DatetimeIndex(datetime_values).is_monotonic_increasing
+
+        nan_columns = datetime_values.isna().any()
+        if nan_columns:
             messages.append(
                 DataCheckError(
-                    message=f"No frequency could be detected in {col_name}, possibly due to uneven intervals.",
+                    message=f"Input datetime column ({col_name}) contains NaN values. Please impute NaN values or drop these rows.",
                     data_check_name=self.name,
-                    message_code=DataCheckMessageCode.DATETIME_HAS_UNEVEN_INTERVALS,
+                    message_code=DataCheckMessageCode.DATETIME_HAS_NAN,
                 ).to_dict()
             )
+            # NaN values are a more significant error than missing values and will break the data check logic, so we return here
+            return messages
+
+        if not inferred_freq:
 
-        if not (pd.DatetimeIndex(datetime_values).is_monotonic_increasing):
+            # Check for only one row per datetime
+            duplicate_dates = pd.Series(datetime_values).diff(1) == pd.Timedelta(0)
+            if duplicate_dates.any():
+                messages.append(
+                    DataCheckError(
+                        message=f"{col_name} has more than one row with the same datetime value.",
+                        data_check_name=self.name,
+                        message_code=DataCheckMessageCode.DATETIME_HAS_REDUNDANT_ROW,
+                    ).to_dict()
+                )
+                # drop any duplicates before continuing
+                datetime_values = datetime_values[~duplicate_dates]
+
+            frequencies = []
+            for i in range(len(datetime_values) - 2):
+                freq = pd.infer_freq(datetime_values[i : i + 3])
+                frequencies.append(freq)
+            num_disruptions = len(set(frequencies))
+
+            # Check for no date missing in ordered dates
+            if num_disruptions > 1 and is_increasing:
+                messages.append(
+                    DataCheckError(
+                        message=f"{col_name} has datetime values missing between start and end date.",
+                        data_check_name=self.name,
+                        message_code=DataCheckMessageCode.DATETIME_IS_MISSING_VALUES,
+                    ).to_dict()
+                )
+
+            # Give a generic uneven interval error if two or more frequencies are detected
+            if num_disruptions > 2 or (num_disruptions > 0 and not is_increasing):
+                messages.append(
+                    DataCheckError(
+                        message=f"No frequency could be detected in {col_name}, possibly due to uneven intervals.",
```

Comment:
If we're going for more information, maybe it might be worth it give the user a place to look.  I know when I was looking at the bike share data, it was irritating having to go through and ID the little "steps" that were occuring over the weekends.  We could also file a follow on issue if you think it's worth doing and worth doing later.

Reply:
The way the logic works out right now, this error will never be the only one raised by this data check - either the `missing_values` or `is_monotonic` check will be there as well. Ideally, these will give more useful information while the uneven error gives the indication that something's significantly wrong with the datetime values. I'm happy to update that in a later issue though.